server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -745,61 +663,104 @@ public final ShardSearchRequest buildShardSearchRequest(SearchShardIterator shar
+                final Semaphore requestPermit = requestPermits.get(shard.getNodeId());
+                assert requestPermit != null;
+                Objects.requireNonNull(requestPermit, "request permit must not be null");
+                requestPermit.release();
+            }
+            if (Thread.currentThread() != executingThread) {
+                start();
+            } else {
+                assert executionLock.availablePermits() == 0;
+                responses.incrementAndGet();
+            }
         }
 
-        Runnable finishAndGetNext() {
-            synchronized (this) {
-                permitsTaken--;
-                assert permitsTaken >= 0 : "illegal taken permits: " + permitsTaken;
+        /**
+         * Called after {@link #performPhaseOnShard(int, SearchShardIterator, SearchShardTarget)} has failed on the target shard.
+         */
+        void onFailure(SearchShardTarget shard, SearchShardIterator shardIt) {
+            assert pendingQueue.contains(shardIt) == false;
+            if (shardIt.remaining() > 0) {
+                pendingQueue.add(shardIt);
             }
-            return tryQueue(null);
+            onSuccess(shard);
         }
 
-        void tryRun(Runnable runnable) {
-            Runnable r = tryQueue(runnable);
-            if (r != null) {
-                r.run();
+        void start() {
+            assert Thread.currentThread() != executingThread;
+            final boolean promised = executionLock.tryAcquire();
+            responses.incrementAndGet();
+            // There's a risk that the executing thread and a responding thread both stop executing shard requests.
+            // We solve this issue by first acquire the execution lock, then increase the response counter, then retry
+            // again if we failed to acquire the lock in the step 1. We have a similar implementation in AsyncIOProcessor
+            // but we do not use it here because it's quite heavy and async.
+            if (promised || executionLock.tryAcquire()) {
+                int processed;
+                do {
+                    processed = responses.get();
+                    try {
+                        executeShardRequests();
+                    } finally {
+                        executionLock.release();
+                    }
+                } while (processed < responses.get() && executionLock.tryAcquire());
             }
         }
 
-        private synchronized Runnable tryQueue(Runnable runnable) {
-            Runnable toExecute = null;
-            if (permitsTaken < permits) {
-                permitsTaken++;
-                toExecute = runnable;
-                if (toExecute == null) { // only poll if we don't have anything to execute
-                    toExecute = queue.poll();
-                }
-                if (toExecute == null) {
-                    permitsTaken--;
+        private void executeShardRequests() {
```

Comment:
It seems costly to have a loop that runs until a semaphore is acquired ? Won't that block the network thread when lots of shards are co-located on the same node and the max shard execution is low ?